csrc/dynamic_transform.cpp

```diff
@@ -0,0 +1,477 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <dynamic_transform.h>
+#include <expr_evaluator.h>
+#include <ir_utils.h>
+#include <lower_utils.h>
+#include <ops/utils.h>
+#include <transform_iter.h>
+#include <transform_view.h>
+#include <utils.h>
+
+#include <optional>
+
+namespace nvfuser {
+
+//! Gather information about concretizing transformations with
+//! concrete input sizes.
+class DynamicTransformInfoBuilder : public IterVisitor {
+ public:
+  DynamicTransformInfoBuilder(Fusion* fusion, ExpressionEvaluator* expr_eval);
+
+  using IterVisitor::handle;
+
+  // Analyze a dynamic reshape and generate AnalyzeViewResult
+  void handle(ViewOp* op) override;
+
+  const auto& getInfo() const {
+    return info_;
+  }
+
+ private:
+  ExpressionEvaluator* expr_eval_ = nullptr;
+
+  DynamicTransformConcretizationInfo info_;
+};
+
+DynamicTransformInfoBuilder::DynamicTransformInfoBuilder(
+    Fusion* fusion,
+    ExpressionEvaluator* expr_eval)
+    : expr_eval_(expr_eval), info_(fusion) {
+  TORCH_INTERNAL_ASSERT(
+      !fusion->isA<kir::Kernel>(),
+      "Invalid container. Kernel container not allowed.\n");
+
+  // Make sure all exactly mapped IDs have the same value in the
+  // evaluator when any one of the IDs has a known value
+  expr_eval_->propagateBoundValuesThroughExactMaps(fusion);
```

Comment:
Yeah, I wonder if the thing to do here is to collapse the internal map instead of mapping many common values. i.e. I just wonder if reducing the number of unique values in the expr evaluator based on these maps could reduce latency for evaluation.

Reply:
Are you thinking about something like `replaceSymbolicSizes`? https://github.com/NVIDIA/Fuser/blob/main/csrc/lower_replace_size.cpp#L161

I think that would be possible. We could do it just once for the original user fusion.